### PR TITLE
Custom stories title

### PR DESCRIPTION
**Pull request by simonrand**
**Background**
We're implementing our components inside an engine and automatically inferring the Stories title from the component class path is starting to make things a bit unwieldy as the number of components grow. 

This PR allows you to override the title of Stories to whatever you wish so you can better manage the grouping/listing of components in Storybook.

### CODE REVIEW

File: lib/view_component/storybook/stories.rb [hunk 1]
```diff
@@ -12,6 +12,10 @@ class Stories
       validate :valid_story_configs
 
       class << self
+        def stories_title(title = nil)
+          self.title = title unless title.nil?
+        end
```
Comment (jonspalmer) [inline lib/view_component/storybook/stories.rb:17]: is there a reason to allow this to be called without any params?

File: spec/dummy/test/components/stories/demo/heading_component_stories.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+module Demo
+  class HeadingComponentStories < ViewComponent::Storybook::Stories
+    stories_title 'Heading Component'
```
Comment (jonspalmer) [inline spec/dummy/test/components/stories/demo/heading_component_stories.rb:5]: Isn't this exactly the value we'd get without this additional feature?

**Review by jonspalmer — COMMENTED** [review]
Thanks for the contribution and especially fixing my terrible spelling. 

I think the fix mostly makes sense - a few minor questions. However, I'd like to understand the problem a bit more. When you say unwieldy what is the challenge? Can you share example?

File: spec/dummy/test/components/stories/demo/heading_component_stories.rb (hunk 2 above)
Comment (simonrand) [reply spec/dummy/test/components/stories/demo/heading_component_stories.rb:5]: In this particular case the title would resolve to `Demo/Heading Component` - see my full comment for more on why this isn't always desirable (in our case)

File: lib/view_component/storybook/stories.rb (hunk 1 above)
Comment (simonrand) [reply lib/view_component/storybook/stories.rb:17]: So actually no – I followed the `def layout` format incorrectly, will update

**Comment by simonrand** [comment]
> Thanks for the contribution and especially fixing my terrible spelling.
> 
> I think the fix mostly makes sense - a few minor questions. However, I'd like to understand the problem a bit more. When you say unwieldy what is the challenge? Can you share example?

Thanks for reviewing!

Sure - for more background, obviously the `title` attribute of the generated story JSON is derived from the class and parent module names, so in the example I added in the specs here, `Demo::HeadingComponentStories` results in a title of `Demo/Heading Component`, then in Storybook the navigation menu is populated as follows:

image

..this is fine for the most part, however we have two issues related to this:

1. We are developing our components in an engine (let's call it `ComponentEngine`) so there is an additional namespace level
2. We have _many_ components

So if we take a (hypothetical) example like a heading component, we might have a component `ComponentEngine::Typography::Headings::HeadingComponent` (and thus a Stories file `ComponentEngine::Typography::Headings::HeadingComponentStories`)

Currently this would resolve to a title of `Component Engine/Typography/Headings/Heading Component` so in Storybook:

image

..which isn't ideal as we add more to our design system.

Perhaps our use case is unique given our naming/categorisation and engine approach, and in reality our problem is probably more about _removing_ module names from the title, which I suppose could be done with a bit more structure using some config flag, but the blocker for that is still being able to override the title somehow, hence submitting this as the change to allow any form of modification to this.

Hope that's a bit clearer - thanks for all the work on this - it's proving very useful!

**Comment by jonspalmer** [comment]
@simonrand Thanks for the helpful background. It strikes me that your use case is much better served with a way of customizing the story title globally where you register a block that takes the Stories object and returns a string. You point out that you have lots of components so with the implementation we have here you going to need to define the title over and over again likely with the same "convention". If we added config option do you think that would solve your problem too?

**Comment by simonrand** [comment]
> @simonrand Thanks for the helpful background. It strikes me that your use case is much better served with a way of customizing the story title globally where you register a block that takes the Stories object and returns a string. You point out that you have lots of components so with the implementation we have here you going to need to define the title over and over again likely with the same "convention". If we added config option do you think that would solve your problem too?

So in our case our team are happy enough to manually add these titles for now - also this doesn't preclude us from adding some helper which can add some convention that suits our needs..

I did think about adding the config option but couldn't decide on a convention for it that would be useful in the wider context of the library - i.e. should the config trim a specific string from the title? A specific number of parent module names to remove? Hence submitting this title override as it felt like an ok "v1", or a starter for some further discussion.

We're using a fork here for now, so there's no urgency on this, but if you think it's a useful feature to include and have some of your own thoughts on the config approach let me know and I'd be happy to do some more on this!

**Comment by jonspalmer** [comment]
@simonrand so here is my sketch of how we'd solve this more generally:

1. Add a new config setting like 'story_title_generator` here: https://github.com/jonspalmer/view_component_storybook/blob/main/lib/view_component/storybook.rb
something like
```ruby
mattr_accessor :story_title_generator, instance_writer: false., default: ->(stories) { stories.stories_name.humanize.titlecase }
```
1. Use that new setting here: https://github.com/jonspalmer/view_component_storybook/blob/main/lib/view_component/storybook/stories.rb#L95

something like 
```ruby
other.title = Storybook.story_title_generator.call(other)
```

Then in your use case in an initializer you could do this

```ruby
config.view_component_storybook.story_title_generator = ->(stories) { stories.stories_name.chomp("ComponentEngine::).humanize.titlecase }
```

Want to take a shot at something like that?

**Comment by jonspalmer** [comment]
@simonrand  thoughts on the api I suggested?

**Comment by simonrand** [comment]
> @simonrand thoughts on the api I suggested?

Hi @jonspalmer, apologies for the delay in replying - yes this looks great! I'm a bit tied up with some other commitments at the moment, but I will aim to update the PR in the coming weeks

File: lib/view_component/storybook/stories.rb [hunk 3]
```diff
@@ -12,6 +12,10 @@ class Stories
       validate :valid_story_configs
 
       class << self
+        def stories_title(title = nil)
+          self.title = title unless title.nil?
```
Comment (jonspalmer) [inline lib/view_component/storybook/stories.rb:16]:
```suggestion
        def title(title = nil)
          self.stories_title = title unless title.nil?
          stories_title
```

**Comment by jonspalmer** [comment]
I merged a slightly refactored version of this branch in #72. I still think the fixes we discussed here are worth of another PR.

**Comment by simonrand** [comment]
@jonspalmer thanks so much for getting to this and for merging #72 - apologies for dropping off radar again - I'll definitely get to this update as soon as I can

**Pull request closed without merging**